server/apps/immich/src/api-v1/user/user.domain.ts

```diff
@@ -0,0 +1,80 @@
+import { UserEntity } from '@app/database/entities/user.entity';
+import { BadRequestException, ForbiddenException, InternalServerErrorException, Logger } from '@nestjs/common';
+import { genSalt, hash } from 'bcrypt';
+import { AuthUserDto } from '../../decorators/auth-user.decorator';
+import { CreateAdminDto, CreateUserDto, CreateUserOauthDto } from './dto/create-user.dto';
+import { IUserRepository } from './user-repository';
+
+export class UserDomain {
+  constructor(private userRepository: IUserRepository) {}
+
+  private async generateSalt(): Promise<string> {
+    return genSalt();
+  }
+
+  private async hashPassword(password: string, salt: string): Promise<string> {
+    return hash(password, salt);
+  }
+
+  async updateUser(authUser: AuthUserDto, user: UserEntity, update: Partial<UserEntity>): Promise<UserEntity> {
+    if (!authUser.isAdmin && (authUser.id !== user.id || user.id != update.id)) {
+      throw new ForbiddenException('You are not allowed to update this user');
+    }
+
+    try {
+      const payload: Partial<UserEntity> = { ...update };
+      if (payload.password) {
+        const salt = await this.generateSalt();
+        payload.salt = salt;
+        payload.password = await this.hashPassword(payload.password, salt);
+      }
+      return this.userRepository.update(user.id, payload);
+    } catch (e) {
+      Logger.error(e, 'Failed to update user info');
+      throw new InternalServerErrorException('Failed to update user info');
+    }
+  }
+
+  async createUser(createUserDto: CreateUserDto | CreateAdminDto | CreateUserOauthDto): Promise<UserEntity> {
```

Comment:
This is part of the business logic, it does more that just requiring the email.
We have 3 ways of creating a user:
- normal user - with mandatory username and email
- admin user - with mandatory username, email and `isAdmin` === true
- normal user from oauth - with mandatory email and oauthId

Reply:
I guess at the end of the day it doesn't matter much, but it seems like these types are misleading / misaligned with the actual business logic we're enforcing (and how we validate it).

The business logic we're actually enforcing in code is:
1. a unique email is required
2. a local admin must exist or isAdmin must be true
3. If you pass a password it will get hashed.

The business logic you're saying we want is 3 different ways to create a user. Either we need to should update the code to actually enforce _that_ business logic (remove non-whitelisted properties from the dtos and validate the actual dto in this method, requiring one complete, valid dto), or I think we should change the types to be more flexible and reflect the business logic we're enforcing in the code.





Now that I look at it, I think there might be a bug - the local admin check

Reply:
What you are saying makes sense and I kind of agree with you, so I started to look at adding a type like you suggested like this:

```typescript
export interface UserCreateParams {
    email: string;
    oauthId?: string;
    password?: string;
    firstName?: string;
    lastName?: string;
  }  
```

Then with TDD mindset I said OK lets write a test that makes having something like `{ oauthId: undefined, password: undefined }` fail, and of course is not possible to write such a test with our approach of testing from the hexagon boundary (i.e. the service).

It would force me to write a test for the `user.core` to test that behaviour.
So as long as the user.core (or any core module) is used by the services and the services are tested it should be fine.
We validate the business logic I described above through the DTO classes in each service. If in the future we add a `isAdmin` property to the OAuthDto then that behaviour would need to be tested from the OauthService, and the user.core would need to be updated if needed.